File: web/admin/src/app/core/services/metrics.service.ts
import { Filter } from '../models/apiFilter';
import { MetricsList } from '../models/metricsList';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';

import { MetricData } from '../models/overallMetrics';
import { Metrics } from '../models/metrics';
import { MetricsCustom } from '../models/metricsCustom';
import { ConfigService } from './config.service';
import { PaginationResult } from '../models/paginatedResult';

@Injectable({
  providedIn: 'root',
})
export class MetricsService {
  constructor(
    private httpClient: HttpClient,
    private configService: ConfigService,
  ) {}

  getTotalForUser(): Observable<Metrics> {
    const url = `${this.configService.config.rorApi}/v1/metrics`;
    return this.httpClient.get<Metrics>(url);
  }

  getTotal(): Observable<Metrics> {
    const url = `${this.configService.config.rorApi}/v1/metrics/total`;
    return this.httpClient.get<Metrics>(url);
  }

  getForDatacenters(): Observable<MetricsList> {
    const url = `${this.configService.config.rorApi}/v1/metrics/datacenters`;
    return this.httpClient.get<MetricsList>(url);
  }

  getForDatacenter(datacenterName: string): Observable<MetricData> {
    const url = `${this.configService.config.rorApi}/v1/metrics/datacenter/${datacenterName}`;
    return this.httpClient.get<MetricData>(url);
  }

  getForWorkspaces(filter: Filter): Observable<PaginationResult<any>> {
    const url = `${this.configService.config.rorApi}/v1/metrics/workspaces/filter`;
    return this.httpClient.post<PaginationResult<any>>(url, filter);
  }

  getForWorkspacesByDatacenter(datacenter: string, filter: Filter): Observable<PaginationResult<any>> {
    const url = `${this.configService.config.rorApi}/v1/metrics/workspaces/datacenter/${datacenter}/filter`;
    return this.httpClient.post<PaginationResult<any>>(url, filter);
  }

  getForWorkspace(workspaceName: string): Observable<any> {
    const url = `${this.configService.config.rorApi}/v1/metrics/workspace/${workspaceName}`;
    return this.httpClient.get<any>(url);
  }

  getForClusters(): Observable<MetricsList> {
    const url = `${this.configService.config.rorApi}/v1/metrics/clusters`;
    return this.httpClient.get<MetricsList>(url);
  }

  getForClustersByWorkspace(workspaceName: string): Observable<MetricsList> {
    const url = `${this.configService.config.rorApi}/v1/metrics/clusters/workspace/${workspaceName}`;
    return this.httpClient.get<MetricsList>(url);
  }

  getForClusterId(clusterId: string): Observable<MetricData> {
    const url = `${this.configService.config.rorApi}/v1/metrics/cluster/${clusterId}`;
    return this.httpClient.get<MetricData>(url);
  }

  getForClusterByProperty(property: string): Observable<MetricsCustom> {
    const url = `${this.configService.config.rorApi}/v1/metrics/custom/cluster/${property}`;
    return this.httpClient.get<MetricsCustom>(url);
  }
}

File: cmd/api/services/metricsService/metrics_service.go
package metricsservice

import (
	"context"
	"errors"

	datacentersrepo "github.com/NorskHelsenett/ror/internal/mongodbrepo/repositories/datacentersRepo"
	metricsrepo "github.com/NorskHelsenett/ror/internal/mongodbrepo/repositories/metricsRepo"
	workspacesrepo "github.com/NorskHelsenett/ror/internal/mongodbrepo/repositories/workspacesRepo"

	"github.com/NorskHelsenett/ror/pkg/config/configconsts"

	"github.com/NorskHelsenett/ror/pkg/apicontracts"

	"github.com/spf13/viper"
	"go.opentelemetry.io/otel"
)

func GetTotal(ctx context.Context) (*apicontracts.MetricsTotal, error) {
	metrics, err := metricsrepo.GetTotal(ctx)
	if err != nil {
		return nil, errors.New("could not get metrics")
	}

	datacentersCount, _ := datacentersrepo.GetTotalCount(ctx)
	workspacesCount, _ := workspacesrepo.GetTotalCount(ctx)

	if metrics == nil {
		metrics = &apicontracts.MetricsTotal{}
	}

	metrics.DatacenterCount = datacentersCount
	metrics.WorkspaceCount = workspacesCount

	return metrics, nil
}

func GetTotalByUser(ctx context.Context) (*apicontracts.MetricsTotal, error) {
	ctx, span := otel.GetTracerProvider().Tracer(viper.GetString(configconsts.TRACER_ID)).Start(ctx, "metricsservice.GetTotalByUser")
	defer span.End()

	ctx, span1 := otel.GetTracerProvider().Tracer(viper.GetString(configconsts.TRACER_ID)).Start(ctx, "metricsrepo.GetTotalByUser")
	defer span1.End()

	metrics, err := metricsrepo.GetTotalByUser(ctx)
	if err != nil {
		return nil, errors.New("could not get metrics by user")
	}

	span1.End()

	ctx, span2 := otel.GetTracerProvider().Tracer(viper.GetString(configconsts.TRACER_ID)).Start(ctx, "datacentersrepo.GetAllByUser")
	defer span2.End()
	datacenters, _ := datacentersrepo.GetAllByUser(ctx)
	span2.End()

	ctx, span3 := otel.GetTracerProvider().Tracer(viper.GetString(configconsts.TRACER_ID)).Start(ctx, "workspacesrepo.GetAllByUser")
	defer span3.End()
	workspaces, _ := workspacesrepo.GetAllByIdentity(ctx)
	span3.End()

	_, span4 := otel.GetTracerProvider().Tracer(viper.GetString(configconsts.TRACER_ID)).Start(ctx, "Return data")
	defer span4.End()
	metrics.DatacenterCount = int64(len(*datacenters))
	metrics.WorkspaceCount = int64(len(*workspaces))
	span4.End()

	return metrics, nil
}

func GetForDatacenters(ctx context.Context) (*apicontracts.MetricList, error) {
	metrics, err := metricsrepo.GetForDatacenters(ctx)
	if err != nil {
		return nil, errors.New("could not get metrics for datacenters")
	}

	return metrics, nil
}

func GetForDatacenterName(ctx context.Context, datacenterName string) (*apicontracts.MetricItem, error) {
	metrics, err := metricsrepo.GetForDatacenterName(ctx, datacenterName)
	if err != nil {
		return nil, errors.New("could not get metrics for datacenters")
	}

	return metrics, nil
}

func GetForWorkspaces(ctx context.Context, filter *apicontracts.Filter) (*apicontracts.PaginatedResult[apicontracts.Metric], error) {
	metrics, err := metricsrepo.GetForWorkspaces(ctx, filter)
	if err != nil {
		return nil, errors.New("could not get metrics for datacenters")
	}

	return metrics, nil
}

func GetForWorkspacesByDatacenter(ctx context.Context, filter *apicontracts.Filter, datacenterName string) (*apicontracts.PaginatedResult[apicontracts.Metric], error) {
	metrics, err := metricsrepo.GetForWorkspacesByDatacenter(ctx, filter, datacenterName)
	if err != nil {
		return nil, errors.New("could not get metrics for datacenters")
	}

	return metrics, nil
}

func GetForWorkspaceName(ctx context.Context, workspaceName string) (*apicontracts.MetricItem, error) {
	metrics, err := metricsrepo.GetForWorkspaceName(ctx, workspaceName)
	if err != nil {
		return nil, errors.New("could not get metrics for workspace")
	}

	return metrics, nil
}

func GetForClusters(ctx context.Context) (*apicontracts.MetricList, error) {
	metrics, err := metricsrepo.GetForClusters(ctx)
	if err != nil {
		return nil, errors.New("could not get metrics for clusters")
	}

	return metrics, nil
}

func GetForClustersByWorkspace(ctx context.Context, workspaceName string) (*apicontracts.MetricList, error) {
	metrics, err := metricsrepo.GetForClustersByWorkspace(ctx, workspaceName)
	if err != nil {
		return nil, errors.New("could not get metrics for clusters")
	}

	return metrics, nil
}

func GetForClusterid(ctx context.Context, clusterId string) (*apicontracts.MetricItem, error) {
	metrics, err := metricsrepo.GetForClusterid(ctx, clusterId)
	if err != nil {
		return nil, errors.New("could not get metrics for clusterid")
	}

	return metrics, nil
}

func ForClustersByProperty(ctx context.Context, property string) (*apicontracts.MetricsCustom, error) {
	metrics, err := metricsrepo.ForClustersByProperty(ctx, property)
	if err != nil {
		return nil, errors.New("could not get metrics for clusterid")
	}

	return metrics, nil
}
